cardano-api/src/Cardano/Api/Typed.hs

```diff
@@ -0,0 +1,1142 @@
+  where
+    ttoken :: AsType a
+    ttoken = proxyToAsType Proxy
+
+
+deserialiseFromTextEnvelope :: HasTextEnvelope a
+                            => AsType a
+                            -> TextEnvelope
+                            -> Either TextEnvelopeError a
+deserialiseFromTextEnvelope ttoken te = do
+    TextView.expectTextViewOfType (textEnvelopeType ttoken) te
+    TextView.decodeFromTextView fromCBOR te
+
+
+data FromSomeType (c :: * -> Constraint) b where
+     FromSomeType :: c a => AsType a -> (a -> b) -> FromSomeType c b
+
+
+deserialiseFromTextEnvelopeAnyOf :: [FromSomeType HasTextEnvelope b]
+                                 -> TextEnvelope
+                                 -> Either TextEnvelopeError b
+deserialiseFromTextEnvelopeAnyOf types te =
+    case List.find matching types of
+      Nothing ->
+        Left (TextView.TextViewTypeError expectedTypes actualType)
+
+      Just (FromSomeType _ttoken f) ->
+        f <$> TextView.decodeFromTextView fromCBOR te
+  where
+    actualType    = TextView.tvType te
+    expectedTypes = [ textEnvelopeType ttoken
+                    | FromSomeType ttoken _f <- types ]
+
+    matching (FromSomeType ttoken _f) = actualType == textEnvelopeType ttoken
+
+
+writeFileTextEnvelope :: HasTextEnvelope a
+                      => FilePath
+                      -> Maybe TextEnvelopeDescr
+                      -> a
+                      -> IO (Either (FileError ()) ())
+writeFileTextEnvelope path mbDescr a =
+    runExceptT $ do
+      handleIOExceptT (FileIOError path) $ BS.writeFile path content
+  where
+    content = TextView.renderTextView (serialiseToTextEnvelope mbDescr a)
+
+
+readFileTextEnvelope :: HasTextEnvelope a
+                     => AsType a
+                     -> FilePath
+                     -> IO (Either (FileError TextEnvelopeError) a)
+readFileTextEnvelope ttoken path =
+    runExceptT $ do
+      content <- handleIOExceptT (FileIOError path) $ BS.readFile path
+      firstExceptT (FileError path) $ hoistEither $ do
+        te <- TextView.parseTextView content
+        deserialiseFromTextEnvelope ttoken te
+
+
+readFileTextEnvelopeAnyOf :: [FromSomeType HasTextEnvelope b]
+                          -> FilePath
+                          -> IO (Either (FileError TextEnvelopeError) b)
+readFileTextEnvelopeAnyOf types path =
+    runExceptT $ do
+      content <- handleIOExceptT (FileIOError path) $ BS.readFile path
+      firstExceptT (FileError path) $ hoistEither $ do
+        te <- TextView.parseTextView content
+        deserialiseFromTextEnvelopeAnyOf types te
+
+
+-- ----------------------------------------------------------------------------
+-- Error reporting
+--
+
+class Show e => Error e where
+
+    displayError :: e -> String
+
+
+-- | The preferred approach is to use 'Except' or 'ExceptT', but you can if
+-- necessary use IO exceptions.
+--
+throwErrorAsException :: Error e => e -> IO a
+throwErrorAsException e = throwIO (ErrorAsException e)
```

Comment:
In what situations will this be necessary?

Reply:
This is for client code that takes a different opinion on error handling from Erik. Totally optional.